cli/command_repository_upgrade.go

```diff
@@ -72,9 +83,130 @@ func (c *commandRepositoryUpgrade) setup(svc advancedAppServices, parent command
 
 	rollbackCmd.Action(svc.directRepositoryWriteAction(c.forceRollbackAction))
 
+	validateCmd := parent.Command("validate", "Validate the upgraded indexes.")
+
+	validateCmd.Action(svc.directRepositoryWriteAction(c.validateAction))
+
 	c.svc = svc
 }
 
+// assign store the info struct in a map that can be used to compare indexes.
+func assign(iif content.Info, i int, m map[content.ID][2]index.Info) {
+	v := m[iif.GetContentID()]
+	v[i] = iif
+	m[iif.GetContentID()] = v
+}
+
+// loadIndexBlobs load index blobs into indexEntries map.
+func loadIndexBlobs(ctx context.Context, indexEntries map[content.ID][2]index.Info, sm *content.SharedManager, which int, indexBlobInfos []content.IndexBlobInfo) error {
+	d := gather.WriteBuffer{}
+
+	for _, indexBlobInfo := range indexBlobInfos {
+		blobID := indexBlobInfo.BlobID
+
+		indexInfos, err := sm.LoadIndexBlob(ctx, blobID, &d)
+		if err != nil {
+			return errors.Wrapf(err, "failed to load index blob with BlobID %s", blobID)
+		}
+
+		for _, indexInfo := range indexInfos {
+			assign(indexInfo, which, indexEntries)
+		}
+	}
+
+	return nil
+}
+
+// validateAction returns an error of the new V1 index blob content does not match the source V0 index blob content.
+// This is used to check that the upgraded index (V1 index) reflects the content of the old V0 index.
+func (c *commandRepositoryUpgrade) validateAction(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+	indexEntries := map[content.ID][2]index.Info{}
+
+	sm := rep.ContentManager().SharedManager
+
+	indexBlobInfos0, _, err := sm.IndexReaderV0().ListIndexBlobInfos(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list index blobs for old index")
+	}
+
+	indexBlobInfos1, _, err := sm.IndexReaderV1().ListIndexBlobInfos(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list index blobs for new index")
+	}
+
+	if len(indexBlobInfos0) == 0 && len(indexBlobInfos1) > 0 {
+		log(ctx).Infof("index has already been upgraded.  Nothing to do")
+		return nil
+	}
+
+	err = loadIndexBlobs(ctx, indexEntries, sm, 0, indexBlobInfos0)
+	if err != nil {
+		return errors.Wrapf(err, "failed to load index entries for v0 index entry")
+	}
+
+	err = loadIndexBlobs(ctx, indexEntries, sm, 1, indexBlobInfos1)
+	if err != nil {
+		return errors.Wrapf(err, "failed to load index entries for new index")
+	}
+
+	for contentID, indexEntryPairs := range indexEntries {
+		if indexEntryPairs[0] == nil || indexEntryPairs[1] == nil {
+			err = errors.Errorf("lop-sided index entries for contentID %q", contentID)
+		} else {
+			err = checkIndexInfo(indexEntryPairs[0], indexEntryPairs[1])
+		}
+
+		if err != nil {
+			break
+		}
```

Comment:
It might be a good idea to dump a minimum number of failures before breaking out. This will help answering questions like "Did we somehow spoil just a single index or more of them ?".

You can add a CLI flag to let the user specify a non-default minimum too. Not required though.